11_FFTBloom/app_resources/image_fft_first_axis.hlsl

```diff
@@ -0,0 +1,156 @@
+	}
+
+	// Util to write values to output buffer in column major order - this ensures coalesced writes
+	void storeColMajor(uint32_t index, NBL_CONST_REF_ARG(complex_t<scalar_t>) firstValue, NBL_CONST_REF_ARG(complex_t<scalar_t>) secondValue)
+	{
+		colMajorAccessor.set(colMajorOffset(2 * glsl::gl_WorkGroupID().x, index), firstValue);
+		colMajorAccessor.set(colMajorOffset(2 * glsl::gl_WorkGroupID().x + 1, index), secondValue);
+	}
+
+	// Once the FFT is done, each thread should write its elements back. Storage is in column-major order because this avoids cache misses when writing.
+	// Channels will be contiguous in buffer memory. We only need to store outputs 0 through Nyquist, since the rest can be recovered via complex conjugation:
+	// see https://en.wikipedia.org/wiki/Discrete_Fourier_transform#DFT_of_real_and_purely_imaginary_signals
+	// FFT unpacking rules explained in the readme
+	// Also, elements 0 and Nyquist fit into a single complex element since they're both real, and it's always thread 0 holding these values
+	template<typename SharedmemAdaptor>
+	void unload(uint32_t channel, SharedmemAdaptor sharedmemAdaptor)
+	{
+		// Storing even elements of NFFT is storing the bitreversed lower half of DFT - see readme
+		for (uint32_t localElementIndex = 0; localElementIndex < ELEMENTS_PER_THREAD; localElementIndex += 2)
+		{
+			// First element of 0th thread has a special packing rule
+			if (!workgroup::SubgroupContiguousIndex() && !localElementIndex)
+			{
+				complex_t<scalar_t> packedZeroNyquistLo = { preloaded[0].real(), preloaded[1].real() };
+				complex_t<scalar_t> packedZeroNyquistHi = { preloaded[0].imag(), preloaded[1].imag() };
+				storeColMajor(0, packedZeroNyquistLo, packedZeroNyquistHi);
+			}
+			else
+			{
+				complex_t<scalar_t> lo = preloaded[localElementIndex];
+				complex_t<scalar_t> hi = getDFTMirror<SharedmemAdaptor>(localElementIndex, sharedmemAdaptor);
+				workgroup::fft::unpack<scalar_t>(lo, hi);
+				// Divide localElementIdx by 2 to keep even elements packed together when writing
+				storeColMajor(_NBL_HLSL_WORKGROUP_SIZE_ * (localElementIndex >> 1) | workgroup::SubgroupContiguousIndex(), lo, hi);
+			}
+		}
+
+	}
+	LegacyBdaAccessor<complex_t<scalar_t> > colMajorAccessor;
+};
+
+void firstAxisFFT()
+{
+	SharedMemoryAccessor sharedmemAccessor;
+	// Set up the memory adaptor
+	using adaptor_t = accessor_adaptors::StructureOfArrays<SharedMemoryAccessor, uint32_t, uint32_t, 1, _NBL_HLSL_WORKGROUP_SIZE_>;
+	adaptor_t sharedmemAdaptor;
+
+	PreloadedFirstAxisAccessor preloadedAccessor;
+	for (uint32_t channel = 0; channel < CHANNELS; channel++)
+	{
+		preloadedAccessor.preload(channel);
+		workgroup::FFT<ELEMENTS_PER_THREAD, false, _NBL_HLSL_WORKGROUP_SIZE_, scalar_t>::template __call(preloadedAccessor, sharedmemAccessor);
+		// Update state after FFT run
+		sharedmemAdaptor.accessor = sharedmemAccessor;
+		preloadedAccessor.unload<adaptor_t>(channel, sharedmemAdaptor);
```

Comment:
preload is probably better hoisted out of the for loop and loading all channels at once, you can however ... store the channels and perform fft individually